File: frontend/src/lib/resources/savingsGoals.ts
import { apiRequest } from '../api'
import { parseDecimal } from '../decimal'
import type { SavingsGoalInput, SavingsGoalRead } from '../types'

type SavingsGoalWire = Omit<SavingsGoalRead, 'target_amount' | 'current_amount'> & {
  target_amount: string | number
  current_amount: string | number
}

function normalize(wire: SavingsGoalWire): SavingsGoalRead {
  return { ...wire, target_amount: parseDecimal(wire.target_amount), current_amount: parseDecimal(wire.current_amount) }
}

export async function listSavingsGoals(signal?: AbortSignal) {
  const wire = await apiRequest<SavingsGoalWire[]>('/savings-goals', {}, signal)
  return wire.map(normalize)
}

export async function createSavingsGoal(payload: SavingsGoalInput, signal?: AbortSignal) {
  const wire = await apiRequest<SavingsGoalWire>('/savings-goals', { method: 'POST', body: JSON.stringify(payload) }, signal)
  return normalize(wire)
}

export async function updateSavingsGoal(id: string, payload: Partial<SavingsGoalInput>, signal?: AbortSignal) {
  const wire = await apiRequest<SavingsGoalWire>(`/savings-goals/${id}`, { method: 'PUT', body: JSON.stringify(payload) }, signal)
  return normalize(wire)
}

export function deleteSavingsGoal(id: string, signal?: AbortSignal) {
  return apiRequest<void>(`/savings-goals/${id}`, { method: 'DELETE' }, signal)
}

File: backend/app/routers/savings_goals.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import get_current_user
from app.models import SavingsGoal, User
from app.schemas import SavingsGoalCreate, SavingsGoalRead, SavingsGoalUpdate

router = APIRouter(prefix="/savings-goals", tags=["savings-goals"])


@router.get("", response_model=list[SavingsGoalRead])
def list_savings_goals(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return (
        db.query(SavingsGoal)
        .filter(SavingsGoal.tenant_id == current_user.tenant_id)
        .order_by(SavingsGoal.created_at.desc())
        .all()
    )


@router.post("", response_model=SavingsGoalRead, status_code=status.HTTP_201_CREATED)
def create_savings_goal(
    payload: SavingsGoalCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    goal = SavingsGoal(
        tenant_id=current_user.tenant_id,
        name=payload.name,
        target_amount=payload.target_amount,
        current_amount=payload.current_amount,
        target_date=payload.target_date,
    )
    db.add(goal)
    db.commit()
    return goal


@router.get("/{goal_id}", response_model=SavingsGoalRead)
def get_savings_goal(
    goal_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    goal = db.query(SavingsGoal).filter(SavingsGoal.id == goal_id, SavingsGoal.tenant_id == current_user.tenant_id).first()
    if not goal:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Savings goal not found")
    return goal


@router.put("/{goal_id}", response_model=SavingsGoalRead)
def update_savings_goal(
    goal_id: str,
    payload: SavingsGoalUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    goal = db.query(SavingsGoal).filter(SavingsGoal.id == goal_id, SavingsGoal.tenant_id == current_user.tenant_id).first()
    if not goal:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Savings goal not found")
    if payload.name is not None:
        goal.name = payload.name
    if payload.target_amount is not None:
        goal.target_amount = payload.target_amount
    if payload.current_amount is not None:
        goal.current_amount = payload.current_amount
    if payload.target_date is not None:
        goal.target_date = payload.target_date
    db.commit()
    return goal


@router.delete("/{goal_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_savings_goal(
    goal_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    goal = db.query(SavingsGoal).filter(SavingsGoal.id == goal_id, SavingsGoal.tenant_id == current_user.tenant_id).first()
    if not goal:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Savings goal not found")
    db.delete(goal)
    db.commit()
